### PR TITLE
Tracing: replace `family` and `title` with `name`

### PR DESCRIPTION
**Pull request by camdencheek**
This removes the "title" argument to `trace.New()`. This was previously only displayed as an attribute on the span, which is confusing since people would often use a span name for this argument since it is called "title." OpenTelemetry does not have the concept of a family and title (these are from `x/net/trace`, which is no longer used), so this is another step towards reconciling our tracing package with OpenTelemetry.

Basically, I went through all uses of `trace.New()` and either named the span `family.title`, removed the title in the case it was an empty string, or converted the title to a set of more structured attributes.

Now, `trace.New()` only takes a context, a span name, and an optional set of attributes.

## Test plan

CI

### CODE REVIEW

File: cmd/searcher/internal/search/search.go [hunk 1]
```diff
@@ -143,7 +142,7 @@ func (s *Service) search(ctx context.Context, p *protocol.Request, sender matchS
 	defer metricRunning.Dec()
 
 	var tr *trace.Trace
-	tr, ctx = trace.New(ctx, "search", fmt.Sprintf("%s@%s", p.Repo, p.Commit))
+	tr, ctx = trace.New(ctx, "search", attribute.String("repo", string(p.Repo)), attribute.String("commit", string(p.Commit)))
```
Comment (jhchabran) [inline cmd/searcher/internal/search/search.go:145]:
```suggestion
	tr, ctx = trace.New(ctx, "search", p.Repo.Attr(), attribute.String("commit", p.Commit.Attr())
```

Comment (jhchabran) [reply cmd/searcher/internal/search/search.go:145]: While checking the code, I discovered that repo and commit have an `Attr()` method, most probably because it standardize the key name, which is different than the one you used here. 

This makes me think that `trace.New` should accept an interface for these 🤔

File: enterprise/cmd/frontend/internal/auth/oauth/middleware.go [hunk 2]
```diff
@@ -34,7 +34,7 @@ func NewMiddleware(db database.DB, serviceType, authPrefix string, isAPIHandler
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		// This span should be manually finished before delegating to the next handler or
 		// redirecting.
-		span, ctx := trace.New(r.Context(), traceFamily, "Middleware.Handle")
+		span, ctx := trace.New(r.Context(), traceFamily)
```
Comment (jhchabran) [inline enterprise/cmd/frontend/internal/auth/oauth/middleware.go:37]: Don't we need to keep the `"Middleware.handle"` here?

File: internal/api/api.go [hunk 3]
```diff
@@ -23,13 +23,17 @@ type RepoID int32
 // Previously, this was called RepoURI.
 type RepoName string
 
-// RepoHashedName is the hashed name of a repo
-type RepoHashedName string
+func (r RepoName) Attr() attribute.KeyValue {
+	return attribute.String("repo", string(r))
```
Comment (jhchabran) [inline internal/api/api.go:27]: Oh wait, it's you who added it 😆

File: internal/api/api.go [hunk 4]
```diff
@@ -23,13 +23,17 @@ type RepoID int32
...
+}
 
 func (r RepoName) Equal(o RepoName) bool {
 	return strings.EqualFold(string(r), string(o))
 }
 
+// RepoHashedName is the hashed name of a repo
+type RepoHashedName string
```
Comment (jhchabran) [inline internal/api/api.go:35]: typing ftw ✨

File: internal/repos/syncer.go [hunk 5]
```diff
@@ -291,7 +292,7 @@ func (s *Syncer) SyncRepo(ctx context.Context, name api.RepoName, background boo
 
 	logger.Debug("SyncRepo started")
 
-	tr, ctx := trace.New(ctx, "Syncer.SyncRepo", string(name))
+	tr, ctx := trace.New(ctx, "Syncer.SyncRepo", attribute.String("repo", string(name)))
```
Comment (jhchabran) [inline internal/repos/syncer.go:295]: Same, `.Attr()`

**Review by jhchabran — DISMISSED** [review]
LGTM! 

Requesting changes just for the few cases where `string(repo)` instead of `repo.Attr()` :) 

cc @bobheadxi you might want to take a look

File: cmd/searcher/internal/search/search.go (hunk 1 above)
Comment (camdencheek) [reply cmd/searcher/internal/search/search.go:145]: Lol, yep. I added that halfway through the PR and forgot to go back and update all the other spots where I constructed it manually 🤦 

> This makes me think that trace.New should accept an interface for these 🤔

Yeah, I think it would be really nice for it to accept things that implement `Attr()` or `Attrs()` natively. I haven't yet figured out an ergonomic interface for it, but I'm definitely on board with the idea 🙂

Comment (camdencheek) [reply cmd/searcher/internal/search/search.go]: I went ahead and pushed a couple commits that just replace all `attribute.String("repo", string(repo))` with `repo.Attr()`. Same for commit.

File: internal/trace/tracer.go [hunk 6]
```diff
@@ -15,26 +15,24 @@ type Tracer struct {
...
 
 	// Set up the split trace.
 	trace := &Trace{
-		family:        family,
+		name:          name,
 		oteltraceSpan: otelSpan,
 	}
 	if parent := TraceFromContext(ctx); parent != nil {
-		trace.family = parent.family + " > " + family
+		trace.name = parent.name + " > " + name
```
Comment (bobheadxi) [inline internal/trace/tracer.go:35]: Do you think this is still necessary? I think trace hierarchy will have this information already, i.e. the parent's trace name is just the name of the parent span, we probably don't need it repeated on every child span

Comment (camdencheek) [reply internal/trace/tracer.go:35]: This is only used for the trace logger. If you don't think that's necessary, we can remove it and simplify a bunch more things 🙂

Comment (bobheadxi) [reply internal/trace/tracer.go:35]: Ah, I guess it comes in handy with scope and all.

Honestly, I think we should remove it - I think oftentimes it ends up just being duplicated scopes (`foo.bar.foo.bar`)

Comment (bobheadxi) [reply internal/trace/tracer.go:35]: If we keep it I think we can just keep the name, and have the name as an attribute `trace.name` on the trace logger

Comment (camdencheek) [reply internal/trace/tracer.go:35]: on it!

Comment (camdencheek) [reply internal/trace/tracer.go]: I went ahead and removed it. The logger scope should cover most of this information

File: internal/trace/logger.go [hunk 7]
```diff
@@ -6,19 +6,12 @@ import (
...
-// Logger will set the TraceContext on l if ctx has one, and also assign the trace
-// family as a scope if a trace family is found. This is an expanded convenience function
-// around l.WithTrace for the common case.
+// Logger will set the TraceContext on l if ctx has one. This is an expanded
+// convenience function around l.WithTrace for the common case.
 //
-// If you already set the family manually on the logger scope, then you might want to use
-// trace.Context(ctx) instead.
+// If you already set the family manually on the logger scope, then you might
+// want to use trace.Context(ctx) instead.
```
Comment (bobheadxi) [inline internal/trace/logger.go:13]: This no longer applies I think:


```suggestion
```

**Review by bobheadxi — APPROVED** [review]
Nice, thank you! 😁

**Review by mucles — APPROVED** [review]
